library/src/main/java/org/apache/apex/malhar/lib/utils/serde/GenericSerde.java

```diff
@@ -36,31 +41,60 @@
 @InterfaceStability.Evolving
 public class GenericSerde<T> implements Serde<T>
 {
-  private transient ThreadLocal<Kryo> kryos = new ThreadLocal<Kryo>()
+  /**
+   * The default GenericSerde use the default class to serde map
+   */
+  public static final GenericSerde DEFAULT = new GenericSerde();
+  private transient Kryo kryo = new Kryo();
+
+  private Class<? extends T> clazz;
+
+  @SuppressWarnings("rawtypes")
+  private Map<Class, Serde> typeToSerde = Maps.newHashMap();
+
+  public <C> void registerSerde(Class<C> type, Serde<C> serde)
   {
-    @Override
-    public Kryo get()
-    {
-      return new Kryo();
-    }
-  };
+    typeToSerde.put(type, serde);
+  }
 
-  private final Class<? extends T> clazz;
+  @SuppressWarnings({ "unchecked", "rawtypes" })
+  public void registerDefaultSerdes()
+  {
+    registerSerde(String.class, new StringSerde());
+    registerSerde(Long.class, new LongSerde());
+    registerSerde(Integer.class, new IntSerde());
+    registerSerde(ImmutablePair.class, new ImmutablePairSerde());
+    registerSerde(Window.TimeWindow.class, new TimeWindowSerde());
+  }
 
   public GenericSerde()
   {
-    this.clazz = null;
+    this(null);
   }
 
   public GenericSerde(Class<? extends T> clazz)
   {
     this.clazz = clazz;
+    registerDefaultSerdes();
+  }
+
+  public Serde getDefaultSerde(Class type)
+  {
+    return typeToSerde.get(type);
   }
 
   @Override
   public void serialize(T object, Output output)
   {
-    Kryo kryo = kryos.get();
+    Class type = object.getClass();
+    if (clazz == null) {
```

Comment:
This is going to be a problem because the clazz may not be known when you deserialize. The clazz member variable, if not null assumes that the class is known during deserialization.